# Home Work Five - ML Zoomcamp 2025

Hi, I'm Norman! This is my submission for Home Work 5 for ML ZoomCamp 2025.

Let’s connect on [LinkedIn](https://www.linkedin.com/in/anormanangel/) and [Twitter](https://x.com/anormanangel) to keep learning together.


## Task

We recommend using python 3.12 or 3.13 in this homework.

In this homework, we're going to continue working with the lead scoring dataset. You don't need the dataset: we will provide the model for you.

In [1]:
!pip install -U email-validator

In [2]:
# Importing the Libraries

import pickle
from fastapi import FastAPI
from pydantic import BaseModel

## Task 1: Installing uv

In [3]:
pip install uv


Note: you may need to restart the kernel to use updated packages.


In [4]:
uv --version


uv 0.9.5 (d5f39331a 2025-10-21)
Note: you may need to restart the kernel to use updated packages.


## Task 2: Installing scikit-learn

In [5]:
uv add scikit-learn==1.6.1


Resolved 6 packages in 4ms
Audited 5 packages in 0.16ms
Note: you may need to restart the kernel to use updated packages.


## Task 3: Finding the properbility of the lead score to convert

In [6]:

# Use curl or wget to download the model file from CMD

# curl -L -o pipeline_v1.bin https://github.com/DataTalksClub/machine-learning-zoomcamp/raw/refs/heads/master/cohorts/2025/05-deployment/pipeline_v1.bin

# wget https://github.com/DataTalksClub/machine-learning-zoomcamp/raw/refs/heads/master/cohorts/2025/05-deployment/pipeline_v1.bin


In [7]:
!pip install wget

In [8]:
# Load the pipeline
with open("pipeline_v1.bin", "rb") as f_in:
    pipeline = pickle.load(f_in)

# Record to score
record = {
    "lead_source": "paid_ads",
    "number_of_courses_viewed": 2,
    "annual_income": 79276.0
}

# Make prediction
probability = pipeline.predict_proba([record])[0, 1]  # probability of positive class

print(f"Probability of lead converting: {probability:.4f}")


Probability of lead converting: 0.5336


In [9]:
# Load the pipeline
with open("pipeline_v1.bin", "rb") as f_in:
    pipeline = pickle.load(f_in)

# Input record
lead = {
    "lead_source": "paid_ads",
    "number_of_courses_viewed": 2,
    "annual_income": 79276.0
}

# Make prediction
prob = pipeline.predict_proba([lead])[0, 1]

print(f"Probability of conversion: {prob:.3f}")

Probability of conversion: 0.534


## Task 4: Finding the properbility of the lead score to get subscription

* What's the probability that this client will get a subscription

Answer: 0.534

In [15]:
# Install Fast API and Uvicorn

!pip install fastapi uvicorn requests


## Steps

* STEP 1: Created main.py in the same directory
Create and Save it as main.py in the same directory as pipeline_v1.bin.


- STEP 2 — Run the app
uvicorn main:app --host 0.0.0.0 --port 9696 --reload
INFO:     Uvicorn running on http://0.0.0.0:9696


- STEP 3: Then open your browser and go to: http://127.0.0.1:9696/docs

There you’ll see the interactive FastAPI interface where you can test your /predict endpoint.

In [11]:
# main.py

# Load the pipeline
with open("pipeline_v1.bin", "rb") as f:
    pipeline = pickle.load(f)

# FastAPI app instance
app = FastAPI()

# Input data model
class Lead(BaseModel):
    lead_source: str
    number_of_courses_viewed: int
    annual_income: float

@app.post("/predict")
def predict(lead: Lead):
    lead_dict = lead.dict()
    probability = pipeline.predict_proba([lead_dict])[0, 1]
    return {"conversion_probability": float(probability)}


In [16]:
uv add fastapi uvicorn

Resolved 21 packages in 35ms
Audited 19 packages in 5ms
Note: you may need to restart the kernel to use updated packages.


## Question 5

Download the base image agrigorev/zoomcamp-model:2025. You can easily make it by using docker pull command.

So what's the size of this base image?

45 MB
121 MB
245 MB
330 MB
You can get this information when running docker images - it'll be in the "SIZE" column.



## Steps

Step 1 — Pull the base image

Run this in your terminal (after Docker Desktop is running):

docker pull agrigorev/zoomcamp-model:2025


This will download the prepared image from Docker Hub.

Step 2 — Check the image size

After the pull finishes, run:

docker images


You’ll see output like this:

REPOSITORY                    TAG       IMAGE ID       CREATED        SIZE
agrigorev/zoomcamp-model      2025      123abc456def   3 weeks ago    121MB


The correct answer is:

121 MB